axum/src/routing/mod.rs

```diff
@@ -697,3 +700,73 @@ fn traits() {
     use crate::test_helpers::*;
     assert_send::<Router<()>>();
 }
+
+/// Trait for things that can provide routes.
+///
+/// Used with [`Router::merge`].
+///
+/// # Example
+///
+/// This can be used to build your own abstractions for building routes and have it work with
+/// [`Router::merge`].
+///
+/// Consider this hypothetical API for having a bunch of paths all go to the same handler:
+///
+/// ```rust
+/// use axum::{
+///     Router,
+///     routing::{get, post, MethodRouter, HasRoutes},
+///     handler::Handler,
+///     body::HttpBody,
+/// };
+///
+/// struct ToSameHandler<B> {
+///     route: MethodRouter<B>,
+///     routes: Router<B>,
+/// }
+///
+/// impl<B> ToSameHandler<B>
+/// where
+///     B: HttpBody + Send + 'static,
+/// {
+///     fn new<H, T>(handler: H) -> Self
+///     where
+///         H: Handler<T, B>,
+///         T: 'static,
+///     {
+///         Self {
+///             route: get(handler),
+///             routes: Router::new(),
+///         }
+///     }
+///
+///     fn add_path(mut self, path: &str) -> Self {
+///         self.routes = self.routes.route(path, self.route.clone());
+///         self
+///     }
+/// }
+///
+/// impl<B> HasRoutes<B> for ToSameHandler<B> {
+///     fn routes(self) -> Router<B> {
+///         self.routes
+///     }
+/// }
+///
+/// let app = Router::new().merge(
+///     ToSameHandler::new(|| async {})
+///         .add_path("/foo")
+///         .add_path("/bar")
+///         .add_path("/baz")
+/// );
+/// # let _: Router<axum::body::Body> = app;
+/// ```
+pub trait HasRoutes<B> {
```

Comment:
Any other suggestions for the name? I don't love `HasRoutes` but have no other ideas 🤔

Reply:
`MakeRouter`? lol

Reply:
That's not too bad actually 🧐 I think I like that more.

Reply:
@jplatte do you have an opinion about the name? Otherwise I'm going with `MakeRouter` I think.

Reply:
Could even use `Into<Router<B>>` instead of a dedicated trait.

Reply:
I don't care too much about the name. I can only add more suggestions, maybe you like one of them even more:

* `IntoRouter`
* `Into<Router>` (i.e. no new trait at all)